### ISSUE TITLE
in attention_rescoring，why the embedding vector of the input token of decoder  multiply self.xscale ?

### ISSUE THREAD

**Issue by l2009312042** [mobvoi/wenet#45]
x = x * self.xscale  in this place   self.xscale  = 16 ? is there any paper explain this ?thanks

**Comment by robin1001** [comment]
The code here is just copied from espnet. I didn't go through the details here. Anyone know about that? @placebokkk @whiteshirt0429

**Comment by placebokkk** [comment]
Please refer to the original transformer paper, 
> Similarly to other sequence transduction models, we use learned embeddings to convert the input
> tokens and output tokens to vectors of dimension d_model. We also use the usual learned linear transformation and softmax function to convert the decoder output to predicted next-token probabilities. In
> our model, we share the same weight matrix between the two embedding layers and the pre-softmax
> linear transformation, similar to [30]. In the embedding layers, we multiply those weights by the square root of d_model.

But I haven't found any reasonable explanation for this scaling in the embedding layer.

Some blogs or discussions just say this scaling is for the sharing weight between the decoder embedding and the decoder pre-softmax liner weights. But I can't see the causality.

And in many implementation, this sharing is not implemented - the pre-softmax linear use a different weight from the embedding one. But the scaling is still added in the code. e.g. I do not see wenet and espnet use this kind of sharing structure. (not very sure)

The explanation in this blog makes a little sense but still lacks proof. see: https://towardsdatascience.com/how-to-code-the-transformer-in-pytorch-24db27c8f9ec

> The reason we increase the embedding values before addition is to make the positional encoding relatively smaller. This means the original meaning in the embedding vector won’t be lost when we add them together.

Maybe it is just a workable magic trick.

Some discussions about this question:
https://stats.stackexchange.com/questions/401402/attention-is-all-you-need-input-scaling-explanation
https://stackoverflow.com/questions/56930821/why-does-embedding-vector-multiplied-by-a-constant-in-transformer-model
https://github.com/OpenNMT/OpenNMT-py/issues/1722
https://github.com/espnet/espnet/issues/2797

**Comment by l2009312042** [comment]
thanks for your reply , the time you reply is about at 2 o'clock in the middle night , thank you for your hard work.

**Comment by placebokkk** [comment]
> thanks for your reply , the time you reply is about at 2 o'clock in the middle night , thank you for your hard work.

np. Thanks for trying Wenet. Welcome any feedback.

**Comment by placebokkk** [comment]
@l2009312042 
And it is better to change 'ctc output embedding vector'  in the issue title to the 'the embedding vector of the input token of deccoder'
